prov/verbs/src/ep_rdm/verbs_rdm_cm.c

```diff
@@ -129,34 +131,70 @@ int fi_ibv_rdm_tagged_repost_receives(struct fi_ibv_rdm_tagged_conn *conn,
 	return num_to_post - rest;
 }
 
-static inline int
+static ssize_t
 fi_ibv_rdm_tagged_prepare_conn_memory(struct fi_ibv_rdm_ep *ep,
 				      struct fi_ibv_rdm_tagged_conn *conn)
 {
+	ssize_t ret = FI_SUCCESS;
 	assert(conn->s_mr == NULL);
 	assert(conn->r_mr == NULL);
 
 	const size_t size = ep->buff_len * ep->n_buffs;
 	conn->s_mr = fi_ibv_rdm_alloc_and_reg(ep,
 				(void **) &conn->sbuf_mem_reg, size);
-	assert(conn->s_mr);
+	if (!conn->s_mr) {
+		assert(conn->s_mr);
+		goto s_err;
+	}
 
 	conn->r_mr = fi_ibv_rdm_alloc_and_reg(ep,
 				(void **) &conn->rbuf_mem_reg, size);
-	assert(conn->r_mr);
+	if (!conn->r_mr) {
+		assert(conn->r_mr);
+		goto r_err;
+	}
 
 	conn->ack_mr = ibv_reg_mr(ep->domain->pd, &conn->sbuf_ack_status,
 		sizeof(conn->sbuf_ack_status),
 		IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE);
 
-	assert(conn->ack_mr);
+	if (!conn->ack_mr) {
+		assert(conn->ack_mr);
+		goto ack_err;
+	}
 
 	conn->rma_mr = fi_ibv_rdm_alloc_and_reg(ep,
 				(void **) &conn->rmabuf_mem_reg, size);
-	assert(conn->rma_mr);
+	if (!conn->rma_mr) {
+		assert(conn->rma_mr);
+		goto rma_err;
+	}
 
 	fi_ibv_rdm_buffer_lists_init(conn, ep);
-	return 0;
+out:
+	return ret;
+
+/* Error handling */
+rma_err:
+	if (conn->rmabuf_mem_reg) {
+		free(conn->rmabuf_mem_reg);
+	}
+ack_err: /* 
+	  * Ack buffer is a part of connection structure, freeing is not needed
+	  */
+r_err:
+	if (conn->rbuf_mem_reg) {
+		free(conn->rbuf_mem_reg);
+	}
+
+s_err:
+	if (conn->sbuf_mem_reg) {
+		free(conn->sbuf_mem_reg);
+	}
```

Comment:
free can accept NULL pointers - you can drop the if statements in the error path